# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [2]:
import pandas as pd #import pandas as pd
#read in the three csv files from the github repository
df_1 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df_2 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv')
df_3 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv')


In [2]:
print(df_1.columns, df_2.columns, df_3.columns) #print the columns of the three dataframes and check if they are the same

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object') Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='object') Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='object')


In [3]:
for df in [df_1, df_2, df_3]: #for each dataframe in the list 
    df.columns = df.columns.str.strip().str.lower() #strip the column names of whitespace and convert them to lowercase
    df.rename(columns={'st': 'state'}, inplace=True) #rename the column 'st' to 'state'
    df.replace(r'^\s+|\s+$', '', regex=True, inplace=True)  #replace leading and trailing whitespace with empty string


In [4]:
df_combined = pd.concat([df_1, df_2, df_3], axis=0) #concatenate the three dataframes into one
df_combined.reset_index(drop=True, inplace=True) #reset the index of the combined dataframe to avoid duplicate indices
df_combined.columns

Index(['customer', 'state', 'gender', 'education', 'customer lifetime value',
       'income', 'monthly premium auto', 'number of open complaints',
       'policy type', 'vehicle class', 'total claim amount'],
      dtype='object')

In [5]:
df_combined.shape # Check the shape of the combined dataframe

(12074, 11)

In [6]:
print(df_combined['gender'].unique()) #Check unique values
df_combined['gender'] = df_combined['gender'].replace({
    'Femal': 'F',
    'female': 'F',
    'Male': 'M',
})
# Replace the values
print(df_combined['gender'].unique()) #Check the new values

[nan 'F' 'M' 'Femal' 'Male' 'female']
[nan 'F' 'M']


In [7]:
print(df_combined['state'].unique()) # Check unique values
df_combined['state'] = df_combined['state'].replace({ #Replace the values
    'WA': 'Washington',
    'AZ': 'Arizona',
    'Cali': 'California',
})
# Check the new values
print(df_combined['state'].unique()) #Check the new values

['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' nan]


In [8]:
print(df_combined['education'].unique()) # Check unique values
df_combined['education'] = df_combined['education'].replace({ #Replace the values
    'Bachelors': 'Bachelor',
    })
df_combined['education'].unique()


['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]


array(['Master', 'Bachelor', 'High School or Below', 'College', 'Doctor',
       nan], dtype=object)

In [9]:
print(df_combined['vehicle class'].unique()) # Check unique values
df_combined['vehicle class'] = df_combined['vehicle class'].replace({ # Replace the values
    'Sports Car': 'Luxury',
    'Luxury SUV': 'Luxury',
    'Luxury Car': 'Luxury',
})
print(df_combined['vehicle class'].unique()) # Check the new values

['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury' nan]


In [10]:
print(df_combined['customer lifetime value']) # Check unique values
# Convert the 'customer lifetime value' column to numeric and round to 2 decimal places
df_combined['customer lifetime value'] = pd.to_numeric(df_combined['customer lifetime value'], errors='coerce').round(2)
print(df_combined['customer lifetime value']) # Check the new values

0                NaN
1         697953.59%
2        1288743.17%
3         764586.18%
4         536307.65%
            ...     
12069    23405.98798
12070    3096.511217
12071    8163.890428
12072    7524.442436
12073    2611.836866
Name: customer lifetime value, Length: 12074, dtype: object
0             NaN
1             NaN
2             NaN
3             NaN
4             NaN
           ...   
12069    23405.99
12070     3096.51
12071     8163.89
12072     7524.44
12073     2611.84
Name: customer lifetime value, Length: 12074, dtype: float64


In [11]:
print(df_combined['number of open complaints'].unique()) # Check unique values
# Safely extract the second value after splitting by '/' and convert to numeric
df_combined['number of open complaints'] = (
	df_combined['number of open complaints']
	.astype(str)
	.str.split('/')
	.str[1]
)
df_combined['number of open complaints'] = pd.to_numeric(df_combined['number of open complaints'], errors='coerce')
print(df_combined['number of open complaints'].unique()) # Check the new values

['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan 0 2 3 1 5 4]
[ 0.  2.  1.  3.  5.  4. nan]


In [12]:
df_combined.shape # Check the shape of the combined dataframe

(12074, 11)

In [13]:
df_combined.isna().sum() #The total number of null values by each collumn 

customer                      2937
state                         2937
gender                        3059
education                     2937
customer lifetime value       5004
income                        2937
monthly premium auto          2937
number of open complaints    10007
policy type                   2937
vehicle class                 2937
total claim amount            2937
dtype: int64

In [14]:
df_combined.dropna(how='all', inplace=True) # Drop rows where all elements are NaN
print(df_combined.isna().sum()) #The total number of null values by each collumn 

customer                        0
state                           0
gender                        122
education                       0
customer lifetime value      2067
income                          0
monthly premium auto            0
number of open complaints    7070
policy type                     0
vehicle class                   0
total claim amount              0
dtype: int64


In [3]:
df_combined.drop(df_combined.drop('number of_open complaints', axis=1, inplace=True)) # Drop the 'number of_open complaints' column


NameError: name 'df_combined' is not defined

In [21]:
df_combined[['gender', 'number of open complaints']] = df_combined[['gender', 'number of open complaints']].fillna('unknown')
# Fill NaN values 

df_combined['customer lifetime value'] = df_combined['customer lifetime value'].fillna(df_combined['customer lifetime value'].median())
print(df_combined.isna().sum()) #The total number of null values by each collumn 

customer                     0
state                        0
gender                       0
education                    0
customer lifetime value      0
income                       0
monthly premium auto         0
number of open complaints    0
policy type                  0
vehicle class                0
total claim amount           0
dtype: int64


In [22]:
numeric_c = df_combined.select_dtypes(include=['number']).columns #convert all the numeric variables to integers
df_combined[numeric_c] = df_combined[numeric_c].astype(int)
df_combined.dtypes

customer                     object
state                        object
gender                       object
education                    object
customer lifetime value       int64
income                        int64
monthly premium auto          int64
number of open complaints    object
policy type                  object
vehicle class                object
total claim amount            int64
dtype: object

In [24]:
print(df_combined.duplicated())# Check for duplicates
print(df_combined.duplicated().sum()) # Check the number of duplicates
print(df_combined.duplicated().any())# Check if there are any duplicates
df_combined.drop_duplicates(inplace=True) # Drop duplicates
print(df_combined.duplicated().sum()) # Check the number of duplicates

0        False
1        False
2        False
3        False
4        False
         ...  
12069    False
12070    False
12071    False
12072    False
12073    False
Length: 9137, dtype: bool
3
True
0


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [13]:
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv') #read csv file from the github repository
df


unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877  California              6857.519928       No   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College        2011-02-18         Employed      M  ...   
1         Basic   College        2011-01-18       Unemployed      F  ...   
2         Basic  Bachelor        2011-02-10         Employed      M  ...   
3      Extended   College        2011-01-11         Employed      M  ...   
4       Premium  Bachelor        2011-01-17    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor        2011-01-19       Unemployed      F  ...   
10906     Basic   College        2011-01-06         Employed      F  ...   
10907  Extended  Bachelor        2011-02-06         Employed      F  ...   
10908   Premium   College        2011-02-13         Employed      F  ...   
10909     Basic  Bachelor        2011-01-08       Unemployed      M  ...   

       number_of_policies     policy_type        policy  renew_offer_type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 27 columns]

In [14]:
print(df.columns) # Check the columns of the dataframe  
df.shape # Check the shape of the dataframe

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type', 'month'],
      dtype='object')


(10910, 27)

In [27]:
if 'unnamed:_0' in df.columns:
	df.drop('unnamed:_0', axis=1, inplace=True) #rewmove the unnamed column if it exists
df.columns = df.columns.str.strip().str.lower() #strip the column names of whitespace and convert them to lowercase
df.columns = df.columns.str.replace('_', ' ', regex=False)  #replace leading and trailing whitespace with empty string
df = df.rename(columns={'employmentstatus': 'employment status'}) #rename the column 'employmentstatus' to 'employment status'
df['education'] = df['education'].replace({
    'Doctor': 'Doctorate',
})
# Replace the values
df.columns # Check the columns of the dataframe

Index(['customer', 'state', 'customer lifetime value', 'response', 'coverage',
       'education', 'effective to date', 'employment status', 'gender',
       'income', 'location code', 'marital status', 'monthly premium auto',
       'months since last claim', 'months since policy inception',
       'number of open complaints', 'number of policies', 'policy type',
       'policy', 'renew offer type', 'sales channel', 'total claim amount',
       'vehicle class', 'vehicle size', 'vehicle type', 'month'],
      dtype='object')

In [16]:
unique_counts = df.nunique() # Count unique values in each column
print(unique_counts)


customer                         9134
state                               5
customer lifetime value          8041
response                            2
coverage                            3
education                           5
effective to date                  59
employmentstatus                    5
gender                              2
income                           5694
location code                       3
marital status                      3
monthly premium auto              202
months since last claim            37
months since policy inception     100
number of open complaints           7
number of policies                  9
policy type                         3
policy                              9
renew offer type                    4
sales channel                       4
total claim amount               5106
vehicle class                       6
vehicle size                        3
vehicle type                        1
month                               2
dtype: int64

In [17]:
print(df['number of policies'].unique()) #Check unique values
print(df['policy'].unique()) 


[9 1 2 7 4 3 6 8 5]
['Corporate L3' 'Personal L3' 'Personal L2' 'Corporate L2' 'Personal L1'
 'Special L1' 'Corporate L1' 'Special L3' 'Special L2']


In [18]:
df.dtypes# Check the data types of the columns

customer                          object
state                             object
customer lifetime value          float64
response                          object
coverage                          object
education                         object
effective to date                 object
employmentstatus                  object
gender                            object
income                             int64
location code                     object
marital status                    object
monthly premium auto               int64
months since last claim          float64
months since policy inception      int64
number of open complaints        float64
number of policies                 int64
policy type                       object
policy                            object
renew offer type                  object
sales channel                     object
total claim amount               float64
vehicle class                     object
vehicle size                      object
vehicle type    

In [19]:
df.isna().sum() #The total number of null values by each collumn 

customer                         0
state                            0
customer lifetime value          0
response                         0
coverage                         0
education                        0
effective to date                0
employmentstatus                 0
gender                           0
income                           0
location code                    0
marital status                   0
monthly premium auto             0
months since last claim          0
months since policy inception    0
number of open complaints        0
number of policies               0
policy type                      0
policy                           0
renew offer type                 0
sales channel                    0
total claim amount               0
vehicle class                    0
vehicle size                     0
vehicle type                     0
month                            0
dtype: int64

In [20]:
print(df.duplicated().sum()) # Check the number of duplicates
print(df.duplicated().any())# Check if there are any duplicates
df.drop_duplicates(inplace=True) # Drop duplicates
print(df.duplicated().sum()) # Check the number of duplicates

443
True
0


In [23]:
pivot_df = df.pivot_table(
    index='sales channel',
    values='total claim amount',
    aggfunc='sum'
).round(2)
print(pivot_df) # Print the pivot table

               total claim amount
sales channel                    
Agent                  1748923.49
Branch                 1259333.44
Call Center             889094.77
Web                     676658.10


Agents generate the highest revenue at 1,748,923, which is more than double that of the Web channel, which contributes the least with only 676,658. The Branchs follows with a revenue of 1,259,333, while the Call Center brings in 889,094. 

It appears that personal services are outperforming digital and technology services, suggesting that the company should consider investing more in the digital and tech sectors.

In [26]:
pivot_df2 = pd.pivot_table(
    df,
    index='gender',
    columns='education',
    values='customer lifetime value',
    aggfunc='mean'
).round(2)

print(pivot_df2)
    

education  Bachelor  College  Doctorate  High School or Below   Master
gender                                                                
F           7913.45  7755.03    7369.59               8616.66  8226.36
M           7713.30  8086.29    7414.86               8162.36  8154.61


Overall, the customer lifetime value is higher for females than for males. Additionally, customers with a high school education or below have the highest average lifetime value, while those with a doctoral degree have the lowest. This suggests that females tend to spend more and remain loyal to the company for a longer period of time. Furthermore, individuals with higher levels of education are more selective compared to those with lower levels of education. The company should create a marketing plan targeting females with low level of education.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [26]:
# Your code goes here